File: services/catalog.go
package services

import (
	"context"
	"log"
	"ng-catalog-service/models"
	"ng-catalog-service/proto"
	"ng-catalog-service/repositories"
)

type CatalogService struct{ proto.CatalogServiceServer }

func (s *CatalogService) Get(ctx context.Context, in *proto.Empty) (*proto.Catalogs, error) {
	log.Println("[*] [CatalogService] Get")

	catalogs, err := repositories.Catalog.Get()
	if err != nil {
		return nil, err
	}

	result := proto.Catalogs{}

	for _, obj := range *catalogs {
		result.Catalogs = append(result.Catalogs, obj.Proto())
	}

	return &result, nil
}

func (s *CatalogService) GetById(ctx context.Context, in *proto.Id) (*proto.Catalog, error) {
	log.Println("[*] [CatalogService] GetById")

	catalog, err := repositories.Catalog.GetById(in.Id)
	if err != nil {
		return nil, err
	}

	result := catalog.Proto()

	return result, nil
}

func (s *CatalogService) Insert(ctx context.Context, in *proto.Catalog) (*proto.Status, error) {
	log.Println("[*] [CatalogService] Insert")

	catalog := models.Catalog{}
	catalog.FromProto(in)

	_, err := repositories.Catalog.Insert(catalog)

	if err != nil {
		return nil, err
	}

	return &proto.Status{Code: 0, Message: "Ok"}, nil
}

func (s *CatalogService) Update(ctx context.Context, in *proto.Catalog) (*proto.Status, error) {
	log.Println("[*] [CatalogService] Update")

	catalog := models.Catalog{}
	catalog.FromProto(in)

	_, err := repositories.Catalog.Update(catalog)

	if err != nil {
		return nil, err
	}

	return &proto.Status{Code: 0, Message: "Ok"}, nil
}

func (s *CatalogService) Delete(ctx context.Context, in *proto.Id) (*proto.Status, error) {
	log.Println("[*] [CatalogService] Delete")

	_, err := repositories.Catalog.Delete(in.Id)

	if err != nil {
		return nil, err
	}

	return &proto.Status{}, nil
}

File: src/controllers/catalog.js
import grpc from 'grpc';
import { grpcCode, RPCtoError } from "../configs/strings";
import { validationResult } from "express-validator";

var proto = grpc.load('./protos/catalog.proto')
var CatalogService = new proto.CatalogService(process.env.CATALOG_SERVICE || "localhost:9020", grpc.credentials.createInsecure())

proto = grpc.load('./protos/account.proto')
var AccountService = new proto.AccountService(process.env.ACCOUNT_SERVICE || "localhost:9000", grpc.credentials.createInsecure())

export class CatalogController {
    static insert (req, res) {
        const checker = validationResult(req);
        if (!checker.isEmpty()) {
            let result = grpcCode.InvalidArgument;
            result.errors = checker.array();

            return res.status(400).json(result);
        }

        AccountService.GetById({id: req.userId}, (err, user) => {
            if (err)
                return res.json(RPCtoError(err));
            
            if (user.profile != 'ADMIN')
                return res.status(401).json(JSON.parse(JSON.stringify(grpcCode.PermissionDenied)));

            let body_request = {
                name: req.body.name,
                items: req.body.items.map(x => {
                    return {
                        name: x.name,
                        item_id: x.item_id,
                        image_url: x.image_url,
                    }
                })
            };
            
            CatalogService.Insert(body_request, (err, result) => {
                if(err)
                    return res.json(RPCtoError(err));
                
                let response = JSON.parse(JSON.stringify(grpcCode.OK));
        
                res.json(response);
            });
        });
    }

    static update (req, res) {
        const checker = validationResult(req);
        if (!checker.isEmpty()) {
            let result = grpcCode.InvalidArgument;
            result.errors = checker.array();

            return res.status(400).json(result);
        }

        AccountService.GetById({id: req.userId}, (err, user) => {
            if (err)
                return res.json(RPCtoError(err));
            
            if (user.profile != 'ADMIN')
                return res.status(401).json(JSON.parse(JSON.stringify(grpcCode.PermissionDenied)));

            let body_request = {
                id: req.params.catalogId,
                name: req.body.name,
                items: req.body.items.map(x => {
                    return {
                        name: x.name,
                        item_id: x.item_id,
                        image_url: x.image_url,
                    }
                })
            };
            
            CatalogService.Update(body_request, (err, result) => {
                if(err)
                    return res.json(RPCtoError(err));
                
                let response = JSON.parse(JSON.stringify(grpcCode.OK));
        
                res.json(response);
            });
        });
    }

    static get (req, res) {
        CatalogService.Get({}, (err, result) => {
            if(err)
                return res.json(RPCtoError(err));
            
            let response = JSON.parse(JSON.stringify(grpcCode.OK));
            response.body = result.catalogs
    
            res.json(response);
        })
    }

    static getById (req, res) {
        const checker = validationResult(req);
        if (!checker.isEmpty()) {
            let result = grpcCode.InvalidArgument;
            result.errors = checker.array();

            return res.status(400).json(result);
        }

        CatalogService.GetById({id: req.params.catalogId}, (err, result) => {
            if(err)
                return res.json(RPCtoError(err));
            
            let response = JSON.parse(JSON.stringify(grpcCode.OK));
            response.body = result
    
            res.json(response);
        });
    }

    static delete (req, res) {
        const checker = validationResult(req);
        if (!checker.isEmpty()) {
            let result = grpcCode.InvalidArgument;
            result.errors = checker.array();

            return res.status(400).json(result);
        }

        AccountService.GetById({id: req.userId}, (err, user) => {
            if (err)
                return res.json(RPCtoError(err));
            
            if (user.profile != 'ADMIN')
                return res.status(401).json(JSON.parse(JSON.stringify(grpcCode.PermissionDenied)));
            
            CatalogService.Delete({id: req.params.catalogId}, (err, result) => {
                if(err)
                    return res.json(RPCtoError(err));
                
                let response = JSON.parse(JSON.stringify(grpcCode.OK));
        
                res.json(response);
            });
        });
    }
}